File: backend/app/middleware/api_logger.py
"""
API Logger Middleware
Logs all API requests and responses to database using background tasks
"""
import time
import json
from fastapi import Request
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.background import BackgroundTask
from datetime import datetime, timezone
from app.core.database import AsyncSessionLocal
from app.models.api_logs import APILogBackend


class APILoggerMiddleware(BaseHTTPMiddleware):
    """Middleware to log all API requests and responses"""
    
    async def dispatch(self, request: Request, call_next):
        # Start timing
        start_time = time.time()
        request_time = datetime.now(timezone.utc)
        
        # Extract request information
        method = request.method
        path = str(request.url.path)
        query_params = str(dict(request.query_params)) if request.query_params else None
        user_ip = request.client.host if request.client else None
        user_agent = request.headers.get("user-agent")
        
        # Extract user info from state (set by auth middleware)
        user_id = None
        username = None
        if hasattr(request.state, "user"):
            user_id = getattr(request.state.user, "id", None)
            username = getattr(request.state.user, "username", None)
        
        # Get request headers (filter sensitive data)
        request_headers = dict(request.headers)
        # Remove sensitive headers
        for sensitive_key in ["authorization", "cookie", "x-api-key"]:
            if sensitive_key in request_headers:
                request_headers[sensitive_key] = "[REDACTED]"
        request_headers_str = json.dumps(request_headers)
        
        # Call the endpoint and get response
        response = await call_next(request)
        
        # Calculate duration
        duration_ms = int((time.time() - start_time) * 1000)
        response_time = datetime.now(timezone.utc)
        
        # Extract response information
        status_code = response.status_code
        
        # Get response headers
        response_headers = dict(response.headers)
        response_headers_str = json.dumps(response_headers)
        
        # Only log if it's a JSON API response (not health checks, static files, etc.)
        should_log = (
            path.startswith("/api/") and
            not path.endswith("/health") and
            not path.endswith("/ready")
        )
        
        if should_log:
            # Add background task to save log AFTER response is sent
            response.background = BackgroundTask(
                self._save_log,
                method=method,
                endpoint=path,
                path=path,
                query_params=query_params,
                request_body=None,
                request_headers=request_headers_str,
                user_id=user_id,
                username=username,
                user_ip=user_ip,
                user_agent=user_agent,
                status_code=status_code,
                response_headers=response_headers_str,
                request_time=request_time,
                response_time=response_time,
                duration_ms=duration_ms,
                error_message=None
            )
        
        return response
    
    async def _save_log(self, **kwargs):
        """Save log entry to database (runs in background after response is sent)"""
        try:
            async with AsyncSessionLocal() as session:
                log_entry = APILogBackend(**kwargs)
                session.add(log_entry)
                await session.commit()
        except Exception as e:
            # Don't let logging errors crash the app
            print(f"Failed to save API log: {str(e)}")

File: frontend/react/src/api/apiLogger.ts
/**
 * API Logger for Frontend
 * Logs all API calls to Backend database
 */
import axios from 'axios';

interface APILogEntry {
  method: string;
  url: string;
  endpoint: string;
  request_body?: string;
  request_headers?: string;
  user_id?: number;
  username?: string;
  session_id?: string;
  origin?: string;
  referer?: string;
  app_source?: string;
  request_type?: string;
  direction?: string;
  status_code?: number;
  response_body?: string;
  response_headers?: string;
  success: boolean;
  request_time: string;
  response_time: string;
  duration_ms: number;
  error_message?: string;
  browser_info?: string;
}

class APILogger {
  private readonly API_URL: string;
  private logQueue: APILogEntry[] = [];
  private isProcessing = false;
  private readonly BATCH_SIZE = 10;
  private readonly FLUSH_INTERVAL = 5000; // 5 seconds

  constructor() {
    this.API_URL = import.meta.env.VITE_API_URL || '';
    
    // Start periodic flush
    setInterval(() => this.flush(), this.FLUSH_INTERVAL);
    
    // Flush on page unload
    window.addEventListener('beforeunload', () => this.flush());
  }

  /**
   * Log an API call
   */
  log(
    method: string,
    url: string,
    requestData?: any,
    requestHeaders?: any,
    responseData?: any,
    responseHeaders?: any,
    statusCode?: number,
    success: boolean = false,
    requestTime: Date = new Date(),
    responseTime: Date = new Date(),
    errorMessage?: string
  ) {
    try {
      // Extract endpoint from URL
      const urlObj = new URL(url, window.location.origin);
      const endpoint = urlObj.pathname;

      // Get user info from localStorage
      const userInfo = this.getUserInfo();

      // Calculate duration
      const duration_ms = responseTime.getTime() - requestTime.getTime();

      // Prepare log entry
      const logEntry: APILogEntry = {
        method: method.toUpperCase(),
        url: url,
        endpoint: endpoint,
        request_body: requestData ? this.safeStringify(requestData) : undefined,
        request_headers: requestHeaders ? this.safeStringify(this.filterHeaders(requestHeaders)) : undefined,
        user_id: userInfo.userId,
        username: userInfo.username,
        session_id: this.getSessionId(),
        origin: window.location.origin,  // e.g., https://ulm-rct.ovu.co.il
        referer: window.location.href,   // Full URL of the page making the request
        app_source: 'ulm-react-web',     // Application identifier
        request_type: 'ui',              // Frontend calls are always UI calls
        direction: 'outbound',           // Frontend makes outbound requests to backend
        status_code: statusCode,
        response_body: responseData ? this.safeStringify(responseData, 500) : undefined,
        response_headers: responseHeaders ? this.safeStringify(responseHeaders) : undefined,
        success: success,
        request_time: requestTime.toISOString(),
        response_time: responseTime.toISOString(),
        duration_ms: duration_ms,
        error_message: errorMessage,
        browser_info: this.getBrowserInfo(),
      };

      // Add to queue
      this.logQueue.push(logEntry);

      // Flush if queue is full
      if (this.logQueue.length >= this.BATCH_SIZE) {
        this.flush();
      }
    } catch (error) {
      // Don't let logging errors affect the application
      console.error('Failed to log API call:', error);
    }
  }

  /**
   * Flush the log queue to the server
   */
  private async flush() {
    if (this.logQueue.length === 0 || this.isProcessing) {
      return;
    }

    this.isProcessing = true;
    const logsToSend = [...this.logQueue];
    this.logQueue = [];

    try {
      // Send logs to backend
      await axios.post(`${this.API_URL}/api/v1/logs/frontend/batch`, {
        logs: logsToSend
      }, {
        headers: {
          'Authorization': `Bearer ${localStorage.getItem('ulm_token')}`,
          'Content-Type': 'application/json',
          'X-App-Source': 'ulm-react-web'  // Identify as UI request (not integration)
        }
      });
    } catch (error) {
      // If failed, put logs back in queue
      this.logQueue.unshift(...logsToSend);
      console.error('Failed to send API logs:', error);
    } finally {
      this.isProcessing = false;
    }
  }

  /**
   * Get user info from localStorage or state
   */
  private getUserInfo(): { userId?: number; username?: string } {
    try {
      // Try to get from token or userInfo in localStorage
      const username = localStorage.getItem('ulm_username');
      const userId = localStorage.getItem('ulm_user_id');
      
      return {
        userId: userId ? parseInt(userId) : undefined,
        username: username || undefined
      };
    } catch {
      return {};
    }
  }

  /**
   * Get or create session ID
   */
  private getSessionId(): string {
    let sessionId = sessionStorage.getItem('ulm_session_id');
    if (!sessionId) {
      sessionId = `session_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
      sessionStorage.setItem('ulm_session_id', sessionId);
    }
    return sessionId;
  }

  /**
   * Get browser info
   */
  private getBrowserInfo(): string {
    return `${navigator.userAgent} | ${window.screen.width}x${window.screen.height}`;
  }

  /**
   * Filter sensitive headers
   */
  private filterHeaders(headers: any): any {
    const filtered = { ...headers };
    const sensitiveKeys = ['authorization', 'cookie', 'x-api-key'];
    
    for (const key of sensitiveKeys) {
      if (filtered[key]) {
        filtered[key] = '[REDACTED]';
      }
    }
    
    return filtered;
  }

  /**
   * Safely stringify with size limit
   */
  private safeStringify(obj: any, maxLength: number = 1000): string {
    try {
      const str = typeof obj === 'string' ? obj : JSON.stringify(obj);
      return str.length > maxLength ? str.substring(0, maxLength) + '...' : str;
    } catch {
      return '<unable to stringify>';
    }
  }
}

// Export singleton instance
export const apiLogger = new APILogger();

